### ISSUE TITLE
Implement data sharing with OpenGL/DirectX/etc.

### ISSUE THREAD

**Issue by c0gent** [cogciprocate/ocl#23]
Collaboration thread containing notes and conversation pertaining to implementation of data sharing between OpenCL and OpenGL/DirectX/etc.

**Comment by c0gent** [comment]
A few things to keep in mind during design:

* User-facing interfaces such as the buffer and image command builder functions must not contain aliased primitive types such as `cl_GLuint`. All `cl_*` types are for internal use and should not be visible on public interfaces from anything in the `core` or `standard` modules. This means it will be necessary to use either their underlying primitive types (`cl_GLUint` -> 'u32`) or create new types in cases where the primitive types are not idiomatic (such as when `cl_GLenum` is used).

(...will add to this list)

**Comment by pravic** [comment]
As for sharing design you may contact the [gfx-rs](https://github.com/gfx-rs/) guys (via [gitter](https://gitter.im/gfx-rs/gfx) for example).

**Comment by c0gent** [comment]
Yes I'm already planning to test and design using glium/glutin so additional interopability with gfx-rs should be no problem. I'll definitely remember to run it by them as soon as it's at least working initially.

**Comment by c0gent** [comment]
@Kerosene2000:
Please see my updated status post above and let me know if I should add anything to it.

Do you have an OpenGL library you would like to focus on testing with? I started experimenting glium but it looks like I may need to hack on it to get it working. A lower level library such as [gl-rs](https://github.com/bjz/gl-rs) or just [GLFW](https://github.com/PistonDevelopers/glfw-rs) might be an easier place to start.

Let me know what you think.

**Comment by Kerosene2000** [comment]
I will explain to you why I want this library working with openGL:
 - I like using GPGPU things on my own programs
 - I want to work on a school project with my favorite program, [particle system project video (FR)](https://cdn.intra.42.fr/video/video/471/low_d__projet__particle_system.mp4#t=182) (1million particles and one gravity point).
I think I can publish a basic version of my project when I have something to show. I will work with `glium`/`glutin` on this project.

[I have already worked](https://github.com/tomaka/glium/pull/1438) (and [here too](https://github.com/tomaka/glium/pull/1437)) on the [`glium` library](https://github.com/tomaka/glium) to make it work with [the *abandoned* `rust-opencl` library](https://github.com/luqmana/rust-opencl) ([the same work as here](https://github.com/luqmana/rust-opencl/pull/86))...

> Please see my updated status post above and let me know if I should add anything to it.

You can add that there is possibility to [acquire (image + buffer)](https://github.com/cogciprocate/ocl/blob/master/src/standard/image.rs#L550) and [release (image + buffer)](https://github.com/cogciprocate/ocl/blob/master/src/standard/buffer.rs#L279) `OpenGL objects`, I don't know...

**Comment by c0gent** [comment]
How did you get a reference to the OpenGL context or the window handle using glium? Something which can be passed to [`clCreateContext`](https://www.khronos.org/registry/cl/sdk/1.2/docs/man/xhtml/clCreateContext.html) as property parameter.

> You can add that there is possibility to acquire (image + buffer) and release (image + buffer) OpenGL objects, I don't know...

Correct me if I'm wrong but we need to link the contexts before we can acquire buffers, right? So this feature won't work yet, correct?

Were you able to get this working using luqmana/rust-opencl?

**Comment by Kerosene2000** [comment]
To link the context OpenGL with OpenCL we first [`clCreateContext](https://gist.github.com/Kerosene2000/e58c5a7048afd31c7600bf7b08f49615#file-cl_context_gl_linking-rs-L94) with properties then we [`clCreateFromGLBuffer`](https://www.khronos.org/registry/cl/sdk/1.2/docs/man/xhtml/clCreateFromGLBuffer.html) the `clBuffer` from an existant GL buffer, **acquire** it, **play** with it and **release** it to `OpenGL`

If you want [some working code in `c++`](https://github.com/Kerosene2000/particle-system/blob/master/src/Particles.cpp#L120), I think you can compile (works on OS X and I have added you (@c0gent) to [my private personnal lib](https://github.com/Kerosene2000/libft)), I'm not, real not, proud of this bad, bad [`C++`project](https://github.com/Kerosene2000/particle-system).

(my `libft` is a lib from my school, it just mandatory to use it everywhere even if it's not optimum :cry:)

**Comment by c0gent** [comment]
I'm familiar with how it works. I should have been more clear...

Have you been able to get this working in Rust? If so how, specifically, are you getting the OpenGL context handle from glium? What method are you calling?

**Comment by Kerosene2000** [comment]
here it is: [`CLContext::with_properties`](https://gitlab.com/Kerosene2000/particle-system-rs/blob/master/src/main.rs#L65).

Haha ! I forget, I use this fork [sebcrozet/rust-opencl](https://github.com/sebcrozet/rust-opencl), `luqmana/rust-opencl` seems abandoned.

**Comment by c0gent** [comment]
No I mean from glium!

I already know the OpenCL side.

I'm asking how to do this in OpenGL!

**Comment by Kerosene2000** [comment]
I edit my post, sorry... I'm a little tired (GMT+2 = 5:16 AM)

**Comment by c0gent** [comment]
I see it now... [CGLGetCurrentContext()](https://gitlab.com/Kerosene2000/particle-system-rs/blob/master/src/main.rs#L90)

**Comment by c0gent** [comment]
Ok, heh... so yeah I understand now :)

**Comment by Kerosene2000** [comment]
[It seems to be apple only](https://www.google.fr/search?q=CGLGetCurrentContext)

**Comment by c0gent** [comment]
Ok, I'll try to figure out how to add this to glium.

**Comment by Kerosene2000** [comment]
[glium-doc/search/currentcontext](http://tomaka.github.io/glium/glium/index.html?search=currentcontext)

**Comment by c0gent** [comment]
Yes and I saw [this](http://tomaka.github.io/glutin/cgl/fn.CGLGetCurrentContext.html) too.

**Comment by Kerosene2000** [comment]
humm, [on cocoa](https://github.com/tomaka/glutin/blob/master/src/api/cocoa/headless.rs#L63)...

**Comment by Kerosene2000** [comment]
Here we are ! [cgl crate](https://github.com/servo/cgl-rs) !
I go to my bed know 😴

**Comment by aginiewicz** [comment]
Hello, I'm looking at various OpenGL and OpenCL solutions for Rust (still new to language, but liking it so far). I noticed, that priority for this dropped today, while Glium also have open issue about CL-GL interop (https://github.com/tomaka/glium/issues/356). I wonder if there was any progress during last 6 months on this? If I understand correctly, what is missing is ability to get current context from glutin or similar?

**Comment by Kerollmops** [comment]
Here is an example of ["How to use Cl with GL in Rust"](https://github.com/Kerollmops/particle-system-rs.git) 😄

**Comment by aginiewicz** [comment]
But it is Mac only, due to cgl dependency, right? Would similar approach work on Win and Lin?

**Comment by Kerollmops** [comment]
ooops, hum I don't really know, I only do it on osx, I don't know if @c0gent can help you...

**Comment by c0gent** [comment]
Relatively little is needed to get it working but I simply haven't had, and won't soon have, time to finish this. There are a few things about it I don't entirely understand. I'd be happy to lay out what I know and to point you in the right direction if you're interested in implementing this. Your desire to see it working is all you'll need.

**Comment by aginiewicz** [comment]
This might be good opportunity to start digging on some real rust project, why not - I'd be happy to help so any directions would be appreciated :)

**Comment by c0gent** [comment]
First question I haven't yet answered: How does one get a reference to the OpenGL context in a cross-platform manner. On Apple it is `CGLGetCurrentContext()`. I haven't looked very deeply into how to do this with glium or any other library. We need a reference to the OpenGL created context, whether it be a pointer or integer type (appears to be platform dependent) that we can give to one of the variants of: [ContextPropertyValue](http://doc.cogciprocate.com/ocl_core/ocl_core/types/structs/enum.ContextPropertyValue.html).

In other words, we need to get something to pass to [clCreateContext](https://www.khronos.org/registry/cl/sdk/1.2/docs/man/xhtml/clCreateContext.html) in order to create an OpenCL context from an OpenGL context. @Kerollmops has already done this for Apple (and hopefully he can fill you in on the details here) but we need a cross-platform way to do it.

I don't know how much you know about this topic. Please let me know how much of my rambling makes sense to you and I'll clarify before I go on. :)

**Comment by aginiewicz** [comment]
Makes perfect sense.

Before researching Rust I used PyOpenCL at work, but dynamic nature of Python makes cross platform GL/CL lot easier, because we can pass value without worrying if this glx/egl/whatever/... I also tried some OpenGL ES with Scala on Android in the past, but stayed with LibGDX as a backend.

From what you said till now, seems it would be easiest to extend Glutin with method to get current context for each supported backend - because it already has most backends I've found while looking around - and some code to unify some of it across the platforms.

**Comment by c0gent** [comment]
You're on exactly the right track.

**Comment by aginiewicz** [comment]
How do you think - making `glutin::GLContext` expose something like `glutin:GLContext::getOCLProperties() -> Result<ocl::core::ContextProperties, SomeError>` (where error could say for example that extensions for sharing context are not available, etc) would be enough for rest of API on the side of OpenCL side just work? Or something that can be converted into `ocl::core::ContextProperties` that is.

`glutin::Window` implements `glutin::GLContext` and we I believe can access `glutin::Window` from `glium::backend::glutin_backend::GlutinFacade`, which implements `glium::backend::Facade` which is in turn returned from `build_glium` method on `DisplayBuilder` (just wrapped in Result). So if it was added to `glutin::GLContext`, we should be able to get OpenCL context properties of window (or of any other context?).

Now I'm not experienced in rust enough to know what approach is best - easiest would be to add `ocl-core` dependency to `glutin`, but this would be unused by most users. So it would have to be optional dependency probably - but then if we add extra method using optional dependency to trait which is always used in abstract way (and we are using dynamic dispatch then), wouldn't it be dangerous in terms of backward compatibility or cause massive api changes due to very fundamental role of GLContext?

Now `ocl::core::ContextProperties` seems to have more than we need, so glutin could define its own `OpenCLContextProperties` and we could add convert from glutin context properties to ocl context properties, but it adds layer of indirection. On the plus side it does not add new dependencies to glutin/glium and seems to require relatively small number of changes (compared to other approaches I can think of).

How do you think - what approach would be most Rust-like?

**Comment by c0gent** [comment]
> How do you think - making glutin::GLContext expose something like glutin:GLContext::getOCLProperties() -> Result<ocl::core::ContextProperties, SomeError> (where error could say for example that extensions for sharing context are not available, etc) would be enough for rest of API on the OpenCL side to just work?

You've got the right idea but it's not necessary for glutin to have any knowledge of OpenCL to accomplish this. All glutin (or any other OpenGL library) has to do is return to return a value that can be passed as a [`clCreateContext` parameter](https://www.khronos.org/registry/cl/sdk/1.2/docs/man/xhtml/clCreateContext.html#parameters) (via [`::create_context`](http://doc.cogciprocate.com/ocl_core/ocl_core/fn.create_context.html)). This value would ideally be wrapped in a Rust `enum` but the specifics don't matter. The only two pieces of information we (ocl) need(s) to know is what sort of platform we're working with (whether it be EGL, CGL, GLX, ...) and the specific reference/pointer/id for the OpenGL context (which will be a single primitive). If these two pieces come in a Rust `enum`, great. If not, fine. We must consume whatever we can get and add an appropriate method or methods to [ContextPropertyValue](http://doc.cogciprocate.com/ocl_core/ocl_core/types/structs/enum.ContextPropertyValue.html) to handle it. Adding a function or method and a new type to glutin *may* be necessary. Adding an ocl dependency to glutin would be in poor taste. An ideal design would be such that neither library depends on the other and instead that the relevant information is communicated using standard, boring, types.

> glutin::Window implements glutin::GLContext and I believe we can access glutin::Window from glium::backend::glutin_backend::GlutinFacade, which implements glium::backend::Facade which is in turn returned from build_glium method on DisplayBuilder (just wrapped in Result). So if it was added to glutin::GLContext, we should be able to get OpenCL context properties of window (or of any other context?).

I'm not sure precisely what you mean here but first off, don't even worry about results and error handling. It's a trivial side issue and I'll take care of that (or show you how to) later. For now just `.unwrap()` everything and we'll sort that out later. 

If you're asking whether or not we can get a context from `glutin::GLContext`, my answer is: I don't think so but I'm not sure exactly. @Kerollmops knows more than I do about it and @tomaka knows more than any of us. Perhaps he'll grace us with his wisdom.

> Now I'm not experienced in rust enough to know what approach is best - easiest would be to add ocl-core dependency to glutin, but this would be unused by most users. So it would have to be optional dependency probably - but then if we add extra method using optional dependency to trait which is always used in abstract way (and we are using dynamic dispatch then I think?), wouldn't it be dangerous in terms of backward compatibility or cause massive api changes due to very fundamental role of GLContext?

Adding any dependencies to glutin is not an option as far as I'm concerned. Even if we did add that dependency it wouldn't create any problems with traits, dynamic dispatch, backward compatibility, or api changes.

I sort of halfway understand your train of thought here and you're exactly right to question all of these things but there are lots of things you're still learning about Rust so you'll have to take my word for it when I say: none of those things are an issue.

> Now ocl::core::ContextProperties seems to have more than we need, so glutin could define its own OpenCLContextProperties and we could add convert from glutin context properties to ocl context properties, but it adds layer of indirection. On the plus side it does not add new dependencies to glutin/glium and seems to require relatively small number of changes (compared to other approaches I can think of).

> How do you think - what approach would be most Rust-like?

The ideal approach, as I mentioned, would be to have a function we could call that returns both the current context platform (EGL, CGL, GLX, ...) and the context identifier (pointer, or whatever). These two values would then be handed to [ContextPropertyValue](http://doc.cogciprocate.com/ocl_core/ocl_core/types/structs/enum.ContextPropertyValue.html) and then to one of the context creation methods ([`::create_context`](http://doc.cogciprocate.com/ocl_core/ocl_core/fn.create_context.html), [`Context::new`](http://doc.cogciprocate.com/ocl/ocl/struct.Context.html#method.new), [`ContextBuilder::properties`](http://doc.cogciprocate.com/ocl/ocl/builders/struct.ContextBuilder.html#method.properties), etc.).

For someone new to Rust you're already remarkably insightful about potential issues and concerns. You'll be an expert in no time.

I'm not the best teacher but I'll be happy to try to answer any other questions you have or point you towards some of the excellent documentation we (the Rust community) have already.

**Comment by aginiewicz** [comment]
> This value would ideally be wrapped in a Rust enum but the specifics don't matter. The only two pieces of information we (ocl) need(s) to know is what sort of platform we're working with (whether it be EGL, CGL, GLX, ...) and the specific reference/pointer/id for the OpenGL context (which will be a single primitive). If these two pieces come in a Rust enum, great. If not, fine.

This is what more or less what I meant by passing only some information and converting it on the ocl side. Isn't one more information (display) needed? I think that if we know all three - context, platform and display we can throw context into properties and from display and platform we know which property to set and to which value out of those platform-dependent ones. It looks to me like this is only place where we need platform dependent information, so pulling in backend dependencies to only get the display if it could be provided seems to be much.

So in short, we get something like this from some unsafe method on glutin side (for example `unsafe fn glutin::Window::shareWithCL() -> SharedContextForCL`) and we are good?
```rust
pub enum SharedContextForCL {
  Egl(context: u32, display: *mut c_void),
  Glx(context: u32, display: *mut c_void),
  Wgl(context: u32, display: *mut c_void),
  Cgl(context: u32, display: *mut c_void),
  NotAvailable
}
```
I think this enum would be hard to define with better types, because of conditional compilation - to get correct types we would have to include all backends at once, or create some proxy types only for this purpose, correct?

**Comment by c0gent** [comment]
I'm really not sure about the role of the 'display' in this scheme. Last time I looked there was very little if any documentation about this. I'm in need of an education on this topic. I suppose the best thing to do is to try it out. You may be the pioneer in this area.

Your `enum SharedContextForCl` concept is right on track. You're right, it would be silly not to use a Rust `enum` to express this. A dependency to `glium` will likely be necessary.

You'll likely be explaining things to me from here on out.

**Comment by aginiewicz** [comment]
Well, I mostly based on [cl_khr_gl_sharing v15](https://www.khronos.org/registry/cl/extensions/khr/cl_khr_gl_sharing.txt) registry entry, i.e. the part:

```text
      * When the CGL binding API is supported, the attribute
	CL_CGL_SHAREGROUP_KHR should be set to a CGLShareGroup handle to
	a CGL share group object.

      * When the EGL binding API is supported, the attribute
	CL_GL_CONTEXT_KHR should be set to an EGLContext handle to an
	OpenGL ES or OpenGL context, and the attribute
	CL_EGL_DISPLAY_KHR should be set to the EGLDisplay handle of the
	display used to create the OpenGL ES or OpenGL context.

      * When the GLX binding API is supported, the attribute
	CL_GL_CONTEXT_KHR should be set to a GLXContext handle to an
	OpenGL context, and the attribute CL_GLX_DISPLAY_KHR should be
	set to the Display handle of the X Window System display used to
	create the OpenGL context.

      * When the WGL binding API is supported, the attribute
	CL_GL_CONTEXT_KHR should be set to an HGLRC handle to an OpenGL
	context, and the attribute CL_WGL_HDC_KHR should be set to the
	HDC handle of the display used to create the OpenGL context.
```
so the `display` above is either CL_CGL_SHAREGROUP_KHR, CL_EGL_DISPLAY_KHR, CL_GLX_DISPLAY_KHR or CL_WGL_HDC_KHR depending on case - in all cases it seems to be represented by `void *` in C.

Seems that MacOS version might require only one parameter after all (no context according to this), would have to check the example code from @Kerollmops as I have no mac experience.

Btw, I also based some of it on [this presentation](http://sa10.idav.ucdavis.edu/docs/sa10-dg-opencl-gl-interop.pdf).

**Comment by c0gent** [comment]
Yep, I've seen that presentation. You're right. There's almost nothing known about this topic but that won't stop us. :)

**Comment by c0gent** [comment]
You've already uncovered most if not all of the clues you need in order to implement this. Please let me know what I can do to help, whether it be some orientation within the `ocl-core` library or with the Rust language itself.

**Comment by aginiewicz** [comment]
I've added method `share_with_opencl` to `GlContext` in https://github.com/aginiewicz/glutin/commit/daf4561f543da17513a44d651f86e30606ae57e3 - only on Linux for now as this is my development platform. Next I will try and check on ocl side if we can use this information, based on https://github.com/tomaka/glutin/blob/master/examples/window.rs (this window from example implements GlContext so we should be able to pull out the context for OpenCL in this simple example) and https://github.com/cogciprocate/ocl/tree/master/examples/gl_interop (but using only glutin instead of glium - as we don't need to do any GL now, only get the context).

Anyway, modified glutin compiles, but probably could be improved a lot. Do you have any suggestions?

**Comment by Kerollmops** [comment]
I'm ok with you, retrieving the [actual openGL context](https://github.com/Kerollmops/particle-system-rs/blob/master/src/main.rs#L90) is really ugly on my project, we really need a crate that do it on all platforms !

**Comment by c0gent** [comment]
Looks good to me so far. As @Kerollmops mentioned in his comments, there are a few existing types in glutin that you can reuse though I'm pretty sure you probably expected that.

I'm interested to see the details how this API will work out when applied to all platforms involved. 

On a side note, it may be useful to have an in-between, 'glue', crate that both ocl and glutin can reference to handle some of the intermediary types. It might help simplify what looks like might be a more complex API than I anticipated. I'll leave that decision to you two.

**Comment by Kerollmops** [comment]
And why not creating a Trait and implementing it for multiple backend ?

**Comment by c0gent** [comment]
Yes that also sounds like a good idea. Perhaps you can help him do that if he's not yet comfortable with traits yet.

**Comment by aginiewicz** [comment]
> And why not creating a Trait and implementing it for multiple backends ?

...

> ... if he's not yet comfortable with traits yet.

It's not that - Rust traits are not much different than a mixture of Scala traits and Go interfaces.

But in this case we know exactly how many possible backends are accepted - because OpenCL can accept only context from EGL, GLX, WGL and CGL. Also, CGL requires different fields - it means that fields would have to be stored on the backend side and implement this trait. Sounds good, but I was looking at downcasting in Rust which would be needed then to get to data and it wasn't that easy (I was pointed to `Any` in `std` - and I don't quite like that solution). In this case at least Scala way would be to define sealed trait and set of case classes - which is exact analogous to Rust enums. I would do similar in Haskell with ADT - mostly because we know all possible answers in front.

> there are a few existing types in glutin that you can reuse though I'm pretty sure you probably expected that.

Yes - we could use that if we define fields inside backend. Sure. But if we define it as sum type at the Api level (which I believe makes sense to avoid downcasting or trait object) - then we have to remember that those types are available only on platform XYZ. So if I enter `x11_dl::xlib::_XDisplay` - which always is a surname to raw pointer - this will compile only on platforms that has this defined - and due to conditional compilation, this will not compile on Windows or Mac or UVW.

**Comment by aginiewicz** [comment]
I've adjusted to some hints from comments in [opencl-interop](https://github.com/aginiewicz/glutin/compare/master...aginiewicz:opencl-interop) branch.

**Comment by c0gent** [comment]
Ok so let me get one thing clear... 

I stated earlier that we only needed two pieces of information to create an Opencl context from OpenGL. I now see that some platforms/situations require three (the display being the third). 

Why don't we create three separate functions within glium, each returning one of those three bits of data. Something like `get_hardware_platform() -> int`, `get_context_ptr() -> *_`, `get_display_id -> [whatever]`. We can then just conditionally compile the four or however many variations with the exact same signatures (whatever those end up being) and consume it and create the appropriate enumerated on our end. Obviously we could just as easily create a single function that returns a 3-tuple too if that is easier.

Is this a viable basic design?

**Comment by aginiewicz** [comment]
Might be not as easy on linux - where we have dynamic dispatch based on two different display servers (X11/Wayland) and two backends (EGL/GLX/Caca). At compile time, we have no idea which one will be selected - it totally runtime dependent. On Linux we know that glx library will be available, but not if it will be used. We even don't know if GL sharing will be available (caca backend might be selected if egl/glx are not functional) - this is a todo, because we would have to check sharing extension availability as well.

I believe current solution passes trough dynamic dispatch.

We could actually get the context and data without even modifications to glutin just getting fields of already defined types in glutin, but we will have to mirror all backend selection logic which changes (and this is a current wip upstream as it seems) so it would be harder to keep up - and if platform selection logic diverges, it would beak things up.

**Comment by c0gent** [comment]
I see. Then I would propose that rather than getting too caught up with worrying about coordinating an elaborate trait object and messing around with types, let's see if we can just distill the bits of information into integers or other primitives and go from there.

**Comment by c0gent** [comment]
There are probably already existing constants within the OpenCL or OpenGL specification that we can use to communicate, for example, the platform type.

**Comment by c0gent** [comment]
Just thinking out loud here. Will the spaghetti stick to the wall? :)

**Comment by aginiewicz** [comment]
> let's see if we can just distill the bits of information into integers or other primitives

I believe that's the raw c_void pointers from proposed enum - in most cases two, for MacOS single one.

> There are probably already existing constants within the OpenCL or OpenGL specification that we can use to communicate, for example, the platform type.

As for OpenCL platform I don't think we can get it that easily. Isn't OpenGL just natively running on gpu that handles current display current screen (like `:0.0`)? MultiGPU OpenGL doesn't seem quite well supported/documented. So we might assume platform is first one that provides a GPU for us. We might be able to get vendor string from OpenGL to eliminate some hybrid setups like Intel+nvidia, but we sill have to look for GPU (and do string parsing... not nice) which currently runs the GL context in case of multiple nvidia gpus on deck.

I know of no simple way to do it other than just try it and let if fail - most examples on the internet just selects the OpenCL platform independently as "default platform" and relies on library user to get correct one.

**Comment by c0gent** [comment]
Sorry I meant 'hardware platform' (GLX, EGL, etc...). There is only one of those per 'CL/GL platform' (yeah confusing).

**Comment by c0gent** [comment]
So yeah let's start by creating a function to return the 'hardware platform' type as an integer.

**Comment by c0gent** [comment]
We can use:

```
pub const CL_GL_CONTEXT_KHR:        cl_context_properties = 0x2008;
pub const CL_EGL_DISPLAY_KHR:       cl_context_properties = 0x2009;
pub const CL_GLX_DISPLAY_KHR:       cl_context_properties = 0x200A;
pub const CL_WGL_HDC_KHR:           cl_context_properties = 0x200B;
pub const CL_CGL_SHAREGROUP_KHR:    cl_context_properties = 0x200C;
```

from: https://github.com/cogciprocate/cl-sys/blob/master/src/cl_gl_h.rs.

**Comment by aginiewicz** [comment]
we can use property name id, yes - this will be nearly the same as what we currently have as variant - so when we have the platform id either as number or enum variant - we would have to add two functions, like `get_shared_context` and `get_shared_display`? Or maybe now - because if those are function, what do we set as return type (as it is platform dependent or will have to be upcasted to raw void pointer).

We might try to use marker traits if it will work, something like:

```rust

type PlatformType = u32; // specific values from cl_gl_h.rs

trait SharedContextData;

struct SharedContext {
    platform: PlatformType,
    data: SharedContextData
}

// on GLX
struct GlxSharedContext {
   context: GLXContext,
   display: GLXDisplay
}

impl SharedContextData for GlxSharedContext;
```

and then on OpenCL side checking PlatformType with some
`if platform == glx` then cast `data as GlxSharedContext`?

But again - on the glutin side, we would have to go just as deep as current try - all the way down to api backend - and there we would have to return `SharedContextData` - and create a function to return the number of backend (which now is comunicated trough variant type).

Or maybe you want to create separate library - where only platform backend is returned as number - and we pull in those dependencies and get the context ourselves on ocl side?

**Comment by c0gent** [comment]
> Or maybe you want to create separate library - where only platform backend is returned as number - and we pull in those dependencies and get the context ourselves on ocl side?

Yes maybe something like this would be the best way to go. What do you think @Kerollmops?

**Comment by aginiewicz** [comment]
but are we able to get only "current" context that way? And if yes, can we be sure that it is same context that glutin is using? GLContext has methods like is_current and make_current - after all we also have to somehow bridge it on the other side sooner or later.

**Comment by c0gent** [comment]
Good questions. I'm afraid I'm no help here.

**Comment by aginiewicz** [comment]
What would be perfect, would probably be a very thin set of exchange data types that both glutin and ocl AND things like glium_sdl2 could base on for GL/CL interop. Just describing the data and traits to be implemented by others? But such things would need to be coordinated with more projects that do same thing (like more glium backend providers, etc).

As a starting point I will try to actually get demo running with current approach. If it would be good enough for glutin-ocl interop, we might try opening pull request and starting discussion with other project maintainers on getting more unified API for it. Then (with proof of concept, even if not a best one) we might be able to gather more insight.

I will play with it again during weekend. Thanks for discussion!

**Comment by Kerollmops** [comment]
Hum... I have a little problem with the `glutin-only` method. I think we need to handle all of this [openGL crates](https://github.com/kud1ing/awesome-rust#graphics). Creating a `Trait` can be a great idea, we need to tell the crates to make available some data about openGL.

Take [glfw_window](https://github.com/PistonDevelopers/glfw_window) for example, this is just a little crate that implement `Window` from [pistoncore-window](https://github.com/PistonDevelopers/piston) and make the glfw library available for piston, we need to follow this kind of work.

**Comment by aginiewicz** [comment]
well, at some point end user has to decide on some lib that creates context/window/display.

I believe it can be on users side (who after all created the window himself) to use same lib to get pointers to what we need in some way. I even think this approach does not need changes on ocl side - I think that now we have exported the stuff on glutin side, we can just make an example that creates window using glutin and gets shared context from glutin. Then we wrap raw pointers inside ocl context properties (still inside example binary code, not inside library) it should cool. This shouldn't be something too hard actually.

To create window using glfw or sdl you have to use glfw or sdl api to create window, so just use their api to get shared context and wrap it in ocl context properties and go on from here.

So I don't think this qualify as a glutin-only way, but glutin-first-because-we-have-to-start-somewhere way :)

If we want some cool way to share, then all those gl packages would have to depend on at least very little opencl exchange library - and this might be not the best thing to do, because most people don't need it. Or am I mistaken here? :)

**Comment by c0gent** [comment]
Yes I see what you mean now.

**Comment by Kerollmops** [comment]
As @c0gent said, just ask openGL libraries to share internal context and window identifier, then let `ocl-team` implement the `Trait` that permit great communication with openCL.

The ugly thing I do in `particle-system-rs` is getting the first `GPU` on this platform admitting this is the openGL used device !

**Comment by c0gent** [comment]
I'm still not sure precisely where you want to use a trait but there are things about the OpenGL side I don't understand so maybe you could clarify or maybe @aginiewicz already knows what you mean.

**Comment by c0gent** [comment]
Ok I think I get it. You want to add methods to the pistoncore-window trait, etc? If that's what you mean we can always add those later but yes I agree.

**Comment by Kerollmops** [comment]
No, I mean something like [gfx_windows_* libraries](https://crates.io/search?q=gfx_windows_), all of theses libs returns generics: `Window`, `Device`, `Factory`, `Color` and `Depth` in relation with the actual window manager/graphic API (OpenGL, DX, Vulkan, Metal...). The gfx_window_ you get depends on the graphic API you use.

**Comment by c0gent** [comment]
I see. Yeah that will be a good idea to add once we get the basics working. Dealing with primitives for now will be easier while we get it up and running, then we can later figure out some nice types and traits to make the api cleaner if we want.

**Comment by c0gent** [comment]
@aginiewicz:

I'm curious to know if you've had any time to work on this and anything new you might have learned. Let me know if there's anything I can do to help.

**Comment by aginiewicz** [comment]
Sorry, nope. Time is only limiting factor right now. I'm a supervisor of 7 engineering theses this semester and the submission deadline is only 7 days away. Work eats all my time now, but I will be back soon :)

Happy New Year by the way!

**Comment by c0gent** [comment]
No problem. Happy New Year :)

**Comment by YaLTeR** [comment]
Hello, perhaps while a good way of doing things is being discussed it's possible to add handling of `GlContextKhr` and `GlxDisplayKhr` to `ContextProperties::to_raw()` in `ocl-core`, along with a setter for `GlxDisplayKhr`, and an appropriate setter in `ContextBuilder`? I did this in a fork and it seems to be enough for GL texture sharing to work with GLX (I have a GL context from SDL so I use that). I might be mistaken though as it's my first time trying to do anything with OpenCL.

**Comment by c0gent** [comment]
It's been a while since I've thought about this issue so my memory is starting to fade a little bit. Could I have a look at your code? Perhaps a pull request?

**Comment by YaLTeR** [comment]
Here:
https://github.com/cogciprocate/ocl-core/compare/master...YaLTeR:glx-sharing?expand=1
+
https://github.com/cogciprocate/ocl/compare/master...YaLTeR:glx-sharing?expand=1#diff-5015932a424cd7e12ceaa4451389415e

**Comment by YaLTeR** [comment]
Should I PR this?

**Comment by c0gent** [comment]
Yeah please do. Give me a few days and I'll have a closer look at it and test it out.

Looks good though :)

**Comment by Nopey** [comment]
tried to use the interop, and I can't get it working. Using sdl2, gl-rs, and ocl master.
gl-rs doesn't interact with ocl.
SDL2 provides some raw pointers, such as SDL_GLContext (const *void)
ocl takes a u32 in its gl_context function, but const *void has 64 bits, so you can't even transmute one to the other.
ocl takes a mut *void in it's glx_display func.
Currently getting an error code -1000 from OpenCL, because I'm shoving a glContext into the glx display function (and thats not right, obv)
but how do i take the const *void and turn it into a u32? am I connecting the wrong dots?
My Repo:
https://github.com/Nopey/ocl_interop_test

**Comment by c0gent** [comment]
It's entirely possible that the type accepted by the [`gl_context`](https://github.com/cogciprocate/ocl-core/blob/master/src/types/structs.rs#L200) function is wrong. As noted [here](https://github.com/cogciprocate/ocl-core/blob/master/src/types/structs.rs#L152), many of the types are guesses and haven't ever been tested or used. If I remember correctly, all I really had to go on was the documentation on the [`clCreateContext`](https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clCreateContext.html) page.

I may have some time to tinker around with this later this week but in the mean while you may want to fork [ocl-core](https://github.com/cogciprocate/ocl-core) and experiment with different types for the [`ContextPropertyValue::GlContextKhr`](https://github.com/cogciprocate/ocl-core/blob/master/src/types/structs.rs#L157) variant (and the `gl_context` method). Let me know if you find anything out.

**Comment by YaLTeR** [comment]
Just a note: I had it working on a 32-bit target so pointers fit into a `u32`.

**Comment by c0gent** [comment]
Well should the type just be a `*mut c_void` then? I can't remember why I made it a u32 in the first place but they were all pretty much guesses. 

What was the source type you were using @YaLTeR?

**Comment by YaLTeR** [comment]
I was using `SDL_GLContext`, so `*mut c_void`.

**Comment by c0gent** [comment]
I'm going to go ahead and change it to that. Thanks :)

**Comment by c0gent** [comment]
@Nopey, please try out https://github.com/cogciprocate/ocl-core/commit/3eaa79aca110f4761229573bcc87ba2557d988a1 by using the latest master of [ocl-core](https://github.com/cogciprocate/ocl-core).

You'll need:
```
[replace]
"ocl-core:0.5.1" = { git = "https://github.com/cogciprocate/ocl-core" }
```
in your Cargo.toml.

**Comment by Nopey** [comment]
Testing that now, and as a side note i found error -1000's definition.
```C
#define CL_INVALID_GL_SHAREGROUP_REFERENCE_KHR  -1000
```
from CL/cl_gl.h

**Comment by Nopey** [comment]
ocl fails to compile with the latest ocl-core, because ocl tries to pass a `u32` instead of a `*mut libc::c_void`
I quickly switched ocl's function to use a `*mut ::ffi::c_void`, 
and tried passing in my SDL_GLContext, and it still comes out with error -1000.

I've been reading up on the interop, and it seems to not require a direct pointer to the gl context, because of the handy dandy functions provided by WGL, GLX, and CGL_Sharegroups can just fetch the pointer (on windows,linux, and apple respectively)
```C++
#ifdef __linux__
  cl_context_properties properties[] = {
      CL_GL_CONTEXT_KHR,
      (cl_context_properties)glXGetCurrentContext(),
      CL_GLX_DISPLAY_KHR,
      (cl_context_properties)glXGetCurrentDisplay(),
      CL_CONTEXT_PLATFORM,
      (cl_context_properties)platform,
      0};
#elif defined _WIN32
  cl_context_properties properties[] = {
      CL_GL_CONTEXT_KHR,
      (cl_context_properties)wglGetCurrentContext(),
      CL_WGL_HDC_KHR,
      (cl_context_properties)wglGetCurrentDC(),
      CL_CONTEXT_PLATFORM,
      (cl_context_properties)platform,
      0};
#elif defined TARGET_OS_MAC
  CGLContextObj glContext = CGLGetCurrentContext();
  CGLShareGroupObj shareGroup = CGLGetShareGroup(glContext);
  cl_context_properties properties[] = {
      CL_CONTEXT_PROPERTY_USE_CGL_SHAREGROUP_APPLE,
      (cl_context_properties)shareGroup, 0};
#else
#error Unsupported Platform
#endif
```
Could this allow ocl to support all GL-backed graphics crates, without any hastle?

**Comment by c0gent** [comment]
Forgot to do 4e1fc00860ee42fc155c67a60443e6f04fb1f162 last night.

As far as precisely what pointers the OpenCL context needs, you know about as much as I do about it. I'll tinker around with it this afternoon and see if I can find you some answers.

**Comment by c0gent** [comment]
@Nopey: https://github.com/cogciprocate/ocl-core/commit/24232c67067ad010e2802e02f0e2779143d7d3c8 seems to have done the trick.  Do a `cargo update` and give it a try.

**Comment by c0gent** [comment]
Alright well as I suspected, creating an OpenCL context isn't actually working, it just wasn't being unwrapped in your test program. Now we need to figure out how to pass a "sharegroup reference" (whatever that is) when creating the OpenCL context.

**Comment by c0gent** [comment]
@YaLTeR can you show us how you are acquiring an OpenGL context handle or just generally how you are using this?

**Comment by YaLTeR** [comment]
I'm actually hooking into a game with an existing context so it wouldn't be very good of an example. The game creates a window with a GL context using SDL, and then I take the context using [SDL_GL_GetCurrentContext](https://wiki.libsdl.org/SDL_GL_GetCurrentContext), cast it to `u32` (which works since I'm working with a 32-bit target) and pass it into `.gl_context()`.

**Comment by c0gent** [comment]
Ok thanks. When you update your ocl you'll no longer need to cast it btw.

**Comment by c0gent** [comment]
Question:
How can we call `glXGetCurrentContext()` or `glXGetCurrentDisplay()` using SDL2? My attempts to load and use function pointers (i.e. `let get_ctx = video_subsystem.gl_get_proc_address("glXGetCurrentContext") as *const unsafe extern "C" fn() -> *mut ();`) have failed (segfaults) though I have no idea what I'm doing.

**Comment by YaLTeR** [comment]
I think those are GLX functions not OpenGL, so OpenGL get proc address functions aren't the correct ones to use. Try linking to GLX.

**Comment by c0gent** [comment]
How might one do that? :) I'm currently using @Nopey's test program which depends on the `sdl2` and `gl` crates. I assume one would need the `glx` crate as well then. 

The other question of course is how can this be set up in a platform-agnostic way...

**Comment by YaLTeR** [comment]
I guess the best solution is `dlopen`ing libglx.so when needed and then `dlsym` and call the function. I'm not so sure about linking to libglx.so on linux because it might not exist if the user has no graphics drivers and it probably doens't exist with Wayland (?).

**Comment by c0gent** [comment]
Yeah so basically we're stuck (as always) trying to get a valid OpenGL context to use. The pointer returned by `sdl2::sys::video::SDL_GL_GetCurrentContext` doesn't work when passed to `ContextBuilder::gl_context` (`CL_GL_CONTEXT_KHR`) or `ContextBuilder::glx_display` (`CL_GLX_DISPLAY_KHR`).

@Nopey: see if you can find out any more about using SDL with OpenCL. Perhaps the context ptr returned by `SDL_GL_GetCurrentContext` is not a true OpenGL context (but may contain one in a field?). Failing that, for a linux-specific way, maybe look into using the `glx` crate to load `glXGetCurrentContext()` and/or `glXGetCurrentDisplay()`. Let me know what you find.

**Comment by Nopey** [comment]
>  Now we need to figure out how to pass a "sharegroup reference" (whatever that is) when creating the OpenCL context.

Sharegroup references are an apple only thing as far as I know, with their CGL brand CL GL interop.
what I learnt from [this pdf](http://sa10.idav.ucdavis.edu/docs/sa10-dg-opencl-gl-interop.pdf) is that creating the properties array needs to be done differently for different systems,
hence the IFDEF ball of code I pasted a post back.
The gl_context is now usable, and throws an error -1000 (invalid sharegroup)
I'm gonna have a hack at porting this IFDEF blob to rust, but I doubt much will come of it

**Comment by Nopey** [comment]
(My plan for the hack)
add a library to the mix, that depends on the `ocl-sys`, `glx` if on linux, `cgl` if on Mac, and might eventually use a `wgl` crate for Windows support if it such a crate is made,
named something like `ocl-gl`
that just defines a function for the developers of a program to use, that is equivelant to that IFDEF blob.
(maybe inline that function, but that's starting to get fancy..)

**Comment by c0gent** [comment]
Keep in mind that the only piece we don't have is the `glXGetCurrentContext()`, `glXGetCurrentDisplay()`, etc. parts. The rest is already implemented so for example...
```
#ifdef __linux__
  cl_context_properties properties[] = {
      CL_GL_CONTEXT_KHR,
      (cl_context_properties)glXGetCurrentContext(),
      CL_GLX_DISPLAY_KHR,
      (cl_context_properties)glXGetCurrentDisplay(),
      CL_CONTEXT_PLATFORM,
      (cl_context_properties)platform,
      0};
#elif defined _WIN32
  cl_context_properties properties[] = {
      CL_GL_CONTEXT_KHR,
      (cl_context_properties)wglGetCurrentContext(),
      CL_WGL_HDC_KHR,
      (cl_context_properties)wglGetCurrentDC(),
      CL_CONTEXT_PLATFORM,
      (cl_context_properties)platform,
      0};
#elif defined TARGET_OS_MAC
  CGLContextObj glContext = CGLGetCurrentContext();
  CGLShareGroupObj shareGroup = CGLGetShareGroup(glContext);
  cl_context_properties properties[] = {
      CL_CONTEXT_PROPERTY_USE_CGL_SHAREGROUP_APPLE,
      (cl_context_properties)shareGroup, 0};
#else
#error Unsupported Platform
#endif
```
translates to (roughly):
```
use ocl::core::ContextProperties;

let mut ctx_props = if cfg!(linux) {
  ContextProperties::new()
    .gl_context(glXGetCurrentContext())
    .glx_display(glXGetCurrentDisplay())
    .platform(platform)
} else if cfg!(windows) {
  ContextProperties::new()
    .gl_context(wglGetCurrentContext())
    .wgl_hdc(wglGetCurrentDC())
    .platform(platform)    
} else if cfg!(macos) {
  let gl_context = CGLGetCurrentContext();
  let share_group = CGLGetShareGroup(gl_context);
  ContextProperties::new()
    .cgl_sharegroup(share_group);
    .platform(platform)
} else {
  panic!("Unsupported platform.");
};

let context = Context::builder()
  .properties(ctx_props)
  .build().unwrap();

```

**Comment by Nopey** [comment]
I've been set back by `glx` not telling the linker to link GLX, and then my manual adding of glx (via build.rs) being unable to find it. (even with LD_LIBRARY_PATH set)
`gl_generator` is supposed to add the proper `link` macros, but thats not ocl's problem. going to go raise an issue at `glx`

**Comment by YaLTeR** [comment]
Sorry, forgot to say: you need _both_ `gl_context()` and `glx_display()` set for this to work (per the OpenCL spec), I set the latter to `glx::GetCurrentDisplay()`.

**Comment by Nopey** [comment]
GLX crate no longer working, using gl_generator to directly link to GLX worked.
[Heres a link](https://github.com/Nopey/rust-ocl-interop)
I'm unable to test Mac, but I'll be testing windows later today and just tested basic linux support. 
WGL_HDC hasn't shown up on [your docs](http://doc.cogciprocate.com/ocl/ocl/builders/struct.ContextProperties.html) yet. (I'm guessing you'll just manually rebuild the docs when this becomes more stable)
Can you add a function for setting the EGL display? that way the interop will also work on android.

**Comment by Nopey** [comment]
I think that `QueCtx` isn't visible, or I just havn't been able to satisfy rustc's need for universal function declerations. 
I need to call `QueCtx::Queue( my proqueue)` to give to the 
Did you want the caller of `from_gl_buffer` to create the `QueCtx` themselves or did you want to make it an automatic cast when they pass in their context or queue?
As a workaround I added QueCtx to the lib.rs `pub use` statement.
ocl-core's [`enqueue_acquire_gl_buffer`](http://doc.cogciprocate.com/ocl_core/ocl_core/fn.enqueue_acquire_gl_buffer.html) function should probably have an ocl frontend, so that developers can avoid ocl-core while using interop

**Comment by c0gent** [comment]
[`::egl_display`](https://github.com/cogciprocate/ocl-core/commit/373d3a443c1ab80953f797d25313e0027d62fb96) and [docs](http://doc.cogciprocate.com/ocl_core/ocl_core/types/structs/struct.ContextProperties.html#method.wgl_hdc) done.

Ok so as for the `QueCtx` stuff... it's just meant as a type that no one uses directly. It's just something that has `From` impls for `Context` and `Queue` just for convenience really (because a context can be resolved from a queue. So you should be able just to pass one or the other of those and be fine. Please explain if I'm misunderstanding you. At a minimum it looks like I need to improve the documentation for that part.

**Comment by Nopey** [comment]
It can cast, I was just doing it wrong when I tried to use implicit casting (I forgot the ampersand)
(Basically, ignore everything that I said about QueCtx because QueCtx is working fine, although you could implement the `from ` trait for ProQue.
Then again, couldn't QueCtx be replaced with an implicit casting from Queue to Context, much like Queue/Context to QueCtx?

Are you going to make a front end for the [`enqueue_acquire_gl_buffer`](http://doc.cogciprocate.com/ocl_core/ocl_core/fn.enqueue_acquire_gl_buffer.html)? (and enqueue_release_gl_buffer)
I think the ideal way to go about that would be a function that takes an array of interop buffers, and then unwraps them to get at the raw cl_mem and passes that to the ocl_core function.
Alternatively, a GLBuffer struct could be written and have some enq_acq() and enq_release() functions.
The current ocl_core enqueue_acquire_gl_buffer function is less powerful than the OpenCL function because it can only enqueue one buffer acquire at a time, so I'm gonna change that and submit a pull request.
I'm gonna go fix the one-buffer issue, and leave deciding

**Comment by Nopey** [comment]
Looking at the code, I now see the `ocl` frontend for enqueue_acquire. you're one step ahead of me!

**Comment by c0gent** [comment]
> although you could implement the from trait for ProQue

Good idea.

> Then again, couldn't QueCtx be replaced with an implicit casting from Queue to Context, much like Queue/Context to QueCtx?

It might be kinda confusing for `Queue` to automatically convert to `Context` in all cases. It might not be obvious what a function really wants if someone was just reading someone else's code. I could be wrong. I'm still not sure that I even really like the whole `QueCtx` thing. I dunno, I could be convinced either way.

> Are you going to make a front end for the enqueue_acquire_gl_buffer?

Yeah I forgot to address that in my earlier post. [There are front ends already](https://docs.rs/ocl/0.14.1/ocl/builders/struct.BufferCmd.html#method.gl_acquire) :) (though I've not used them myself).

> The current ocl_core enqueue_acquire_gl_buffer function is less powerful than the OpenCL function because it can only enqueue one buffer acquire at a time, so I'm gonna change that and submit a pull request.

Yes please do that would be great.

**Comment by c0gent** [comment]
I didn't design `enqueue_acquire_gl_buffer` and `enqueue_release_gl_buffer`. I see what you mean about them being less powerful than they could be.

I think the best thing to do is simply expose direct versions of [clEnqueueAcquireGLObjects](https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clEnqueueAcquireGLObjects.html) and [clEnqueueReleaseGLObjects](https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clEnqueueReleaseGLObjects.html) (named accordingly) that accept a vector of [`Mem`](https://docs.rs/ocl-core/0.5.1/ocl_core/types/abs/struct.Mem.html)s.

**Comment by Nopey** [comment]
Here is [the pull request](https://github.com/cogciprocate/ocl-core/pull/9) of the multi-buffer acquire/release in ocl_core.
I'm gonna fix some of the docs for ocl's BufferCmd now, because I noticed odd things in it..
Broken example code, that uses depricated function (should be `.enew(&mut empty_event)`, and not use `read_async`)
```rust
// Reads without blocking:
buffer.cmd().read_async(&dst_vec).enew(&empty_event).enq().unwrap();
```
conflicting directions (the `read()` one is incorrect)
`read()`
> After calling this method, the blocking state of this command will be locked to true and a call to ::block will cause a panic.

`read_async()`
> "Deprecated since 0.13.0
> : Use '::read' with '::block(false)' for unsafe asynchronous reads. "

**Comment by c0gent** [comment]
> Here is the pull request of the multi-buffer acquire/release in ocl_core.

Great :)

> I'm gonna fix some of the docs for ocl's BufferCmd tomorrow, because I noticed odd things in it..
Broken example code, that uses depricated function (should be .enew(&mut empty_event), and not use read_async)
...

Can you elaborate on what you mean here? I'm not quite following you. A [link](https://github.com/cogciprocate/ocl/blob/master/src/lib.rs#L3) to the code you're referring to would be great too. :)

**Comment by Nopey** [comment]
Heres the [BufferCmd Docs](http://doc.cogciprocate.com/ocl/ocl/builders/struct.BufferCmd.html), example code is right at the top, and the `read()` and `read_async()` functions are just below.
The changes I was planning on making were removing read's panic warning, and changing the example code's `.read_async(&dst_vec).enew(&empty_event)` with `.read(&dst_vec).block(false).enew(&mut empty_event)

**Comment by Nopey** [comment]
The code where `.read()`'s `.enq()` is implemented is at line `774`.
It uses the cmd's blocking bool, and the `block` function (line 972) doesn't care if the bool has been set, it just sets it.

**Comment by Nopey** [comment]
Pull request #76 contains my corrections to the documentation

**Comment by Nopey** [comment]
I can't put [`ocl-interp`](https://github.com/Nopey/rust-ocl-interop) on crates.io until ocl updates on crates.io, because of the change from u32 to ptr in `gl_context`
Before a new version of ocl is pushed to crates, I think more testing should happen on the interop.
p.s. EGL (android) support is now in [`ocl-interop`](https://github.com/Nopey/rust-ocl-interop)
I'm writing tests in the interop with various window creation methods on linux, and if my windows installation can boot I'll also test it there.

**Comment by Nopey** [comment]
Basic interop tests written for GLFW, Glutin, and SDL2.
All are passing on my Ubuntu 17.04, about to reboot and set up rust in windows to test them there.
The tests creates a window, then a GL Buffer, then a CL Buffer, then runs a kernel on the CL buffer, outputting data onto the GL Buffer, then copies the GL Buffer's results to a local vec and ensures that the data was computed correctly.
Tests are in the master branch of [`ocl-interop`](https://github.com/Nopey/rust-ocl-gl), so please go and try them out!

These tests do not currently use the GL Buffer in any GL Operations, which is kind of important. Testing [`from_gl_texture` and `from_gl_renderbuffer`](http://doc.cogciprocate.com/ocl/ocl/struct.Image.html#method.from_gl_texture) is also planned, but won't happen for a while.

**Comment by Nopey** [comment]
I can't test for windows because I'm not spending any more time debugging cargo. (after building, when cargo runs the test executible, `cargo.exe:?: can't find the shared library.`)
Is anyone else able to test [`ocl-interop`](http://github.com/Nopey/rust-ocl-interop) on Windows?

**Comment by c0gent** [comment]
> I can't put ocl-interp on crates.io until ocl updates on crates.io, because of the change from u32 to ptr in gl_context

You could put
```
[replace]
"ocl-core:0.5.1" = { git = "https://github.com/cogciprocate/ocl-core" }
"ocl:0.14.1" = { git = "https://github.com/cogciprocate/ocl" }
```
and ask your crate users to do the same but I agree, just hold off on publishing for a while.

> Basic interop tests written for GLFW, Glutin, and SDL2.
All are passing on my Ubuntu 17.04, about to reboot and set up rust in windows to test them there.
The tests creates a window, then a GL Buffer, then a CL Buffer, then runs a kernel on the CL buffer, outputting data onto the GL Buffer, then copies the GL Buffer's results to a local vec and ensures that the data was computed correctly.
Tests are in the master branch of ocl-interop, so please go and try them out!

Great, nice work! I'll play around with it in the next couple of days.

> I can't test for windows because I'm not spending any more time debugging cargo. (after building, when cargo runs the test executible, cargo.exe:?: can't find the shared library.)
Is anyone else able to test ocl-interop on Windows?

I'm guessing it's having trouble finding `OpenCL.lib`. If so, first try using the latest master of `cl-sys` (from the repo, using a `[replace]` line as above). It should find it if it's in any of the usual places. If that doesn't work, first let me know so we can improve the build script that locates it, then find and copy that file to your build target directory. You can also try copying it to `{Rust/Cargo path}\lib\rustlib\{Your rust build, i.e. x68_64-pc-windows-msvc}\lib`.

**Comment by Nopey** [comment]
Thanks for the advice.
I Installed NVidia's SDK in a non-standard location, which did cause OpenCL.lib to fail to link,
and after I added the SDK Lib directory to the LIBRARY env variable, the build completed.
I can build, but when I run cargo test, _cargo.exe_ has a library problem. I don't really care about getting rust running on my windows install, but I'll try copying OpenCL.lib to cargo's lib directory.

Before I do that, I'll write a test that actually uses the gl buffer as a vertex buffer.

**Comment by c0gent** [comment]
Have you done this with the newest `cl-sys` on master? What is the exact cargo error message?

**Comment by Nopey** [comment]
the exact cargo error is `cargo.exe:?: can't find the shared library`

**Comment by c0gent** [comment]
Is OpenCL.lib in your `PATH`?

**Comment by Nopey** [comment]
after some testing, i was unable to get glutin to draw anything from a VBO, but glfw worked fine.
I havn't tried using a VBO in SDL2 yet, because I want to keep the old multiply_by_scalar test around.
I'm rebooting back to windows, and will try what you suggested.

**Comment by Nopey** [comment]
Adding OpenCL.lib and SDL2.lib to my path worked.
forgot to push VBO to master, rebooting twice to try out VBOs on windows

**Comment by Nopey** [comment]
`glutin` able to draw the VBO on windows, not on linux. doesn't seem to be an issue with `ocl-interop`, so I'm just going to ignore it and focus on testing `glium`

**Comment by c0gent** [comment]
I'm getting ready to test it a little as well. Gotta dust off Windows...

**Comment by c0gent** [comment]
So on Linux I get my usual `CL_INVALID_GL_SHAREGROUP_REFERENCE_KHR` error, which is all I ever have got when trying to do this even though my platforms and devices all have the `cl_khr_gl_sharing` extension.

I assume it's working for you on Linux so I wonder what's different about your setup.

**Comment by c0gent** [comment]
Besides the [build script issue](https://github.com/Nopey/rust-ocl-interop/issues/2) (which would be a huge time saver for future users) all tests work on Windows. Not sure what to do on Linux to get it working.

**Comment by Nopey** [comment]
Poking around at this again, I'll let you know how using it goes.
Just realized this could be feature gated inside of ocl, instead of being a self-contained crate.
This would allow for some convenience functions
```rust
let mut context = ocl::Context::builder().properties(ocl-gl::get_properties_list().platform(platform)).build().unwrap();
//could become something like:
let mut context = ocl::Context::builder().enable_interop().build().unwrap()
``

**Comment by c0gent** [comment]
I think that's a great idea. I had imagined integrating it into ocl at some point as well.

Also, as I mentioned at the bottom of [this issue](https://github.com/Nopey/rust-ocl-interop/issues/1), I'm stuck on getting [`get_gl_context_info_khr`](https://github.com/cogciprocate/ocl-core/blob/master/src/functions.rs#L802) to work (in order to get a device id).
